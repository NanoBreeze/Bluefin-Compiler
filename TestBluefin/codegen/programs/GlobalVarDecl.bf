/*
Verifies global variable declaration, including ones without assignemnt, 
compile-time expr assignment, casting, and dynamic assignment
*/

int foo() {
	return 3;
}

int a;			// no expr
int b = 7;		// primaryId expr
int c = 8 + 9 * 2; // larger expr, requires constant folding

void hi(int a) { // check if global vars can be defined after funcDefs
	a + 1;
}

int d = foo(); // requires a special llvm.global_ctor for this function call entry
int e = foo() + 9;
bool f;
float g;  
float h = 5;  // 5 will need to be casted
float i = 5 + foo();  // 5 and 'foo' will need to be casted

float j = e = 4*foo();  // The subexpr 'e = 4' will need to be casted into a float
