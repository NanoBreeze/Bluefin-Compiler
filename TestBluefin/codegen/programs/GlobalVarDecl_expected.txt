; ModuleID = 'GlobalVarDecl.bf'
source_filename = "GlobalVarDecl.bf"

@a = global i32 0
@b = global i32 7
@c = global i32 26
@d = global i32 0
@e = global i32 0
@f = global i1 false
@g = global float 0.000000e+00
@h = global float 5.000000e+00
@i = global float 0.000000e+00
@llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 0, void ()* @_VarDeclInitializations, i8* null }]

define i32 @foo() {
entry:
  ret i32 3
}

define void @hi(i32 %a) {
entry:
  %addtmp = add i32 %a, 1
  ret void
}

define internal void @internalGlobalVarDecl_d() {
entry:
  %calltmp = call i32 @foo()
  store i32 %calltmp, i32* @d, align 4
  ret void
}

define internal void @internalGlobalVarDecl_e() {
entry:
  %calltmp = call i32 @foo()
  %addtmp = add i32 %calltmp, 9
  store i32 %addtmp, i32* @e, align 4
  ret void
}

define internal void @internalGlobalVarDecl_i() {
entry:
  %calltmp = call i32 @foo()
  %addtmp = add i32 5, %calltmp
  %casttmp = sitofp i32 %addtmp to float
  store float %casttmp, float* @i, align 4
  ret void
}

define internal void @_VarDeclInitializations() {
entry:
  call void @internalGlobalVarDecl_d()
  call void @internalGlobalVarDecl_e()
  call void @internalGlobalVarDecl_i()
  ret void
}
